### PR TITLE
Migrate to C# 7

### PR DESCRIPTION
**Pull request by smoogipoo**
Have converted most new resharper suggestions to hints, except `JoinNullCheckWithUsage` which results in weird code in some scenarios. This would transform code in the following way:
```
if (x == null) throw new ArgumentNullException(nameof(x));
this.x = x;
```
->
```
this.x = x ?? throw new ArgumentNullException(nameof(x));
```

We can re-enable these as we see fit in further PRs (especially the out variable one I think).

### CODE REVIEW

**Comment by peppy** [comment]
I personally like the null joining

**Comment by smoogipoo** [comment]
It's okay until you get to situations where you use parameters later in the ctor instead of immediately storing them locally, then it looks like you're doing parameter validation late.

I wouldn't mind leaving it as a hint though.

File: osu.Framework/osu.Framework.csproj [hunk 1]
```diff
@@ -1,7 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
-  <Import Project="$(SolutionDir)\packages\Microsoft.Net.Compilers.2.3.2\build\Microsoft.Net.Compilers.props" Condition="Exists('$(SolutionDir)\packages\Microsoft.Net.Compilers.2.3.2\build\Microsoft.Net.Compilers.props')" />
```
Comment (peppy) [inline osu.Framework/osu.Framework.csproj]: these aren't required any more?

Comment (smoogipoo) [reply osu.Framework/osu.Framework.csproj]: Moved to LangVer.props.

**Pull request merged**